colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -0,0 +1,854 @@
+                ################
+                # chunk = 1 && is_last_stage
+                # do nothing; Already send input_tensor_grad to local_send_bwd_buffer in schedule b;
+                ################
+                if self.stage_manager.is_last_stage(ignore_chunk=True):
+                    return []
+
+                ################
+                # chunk = 1 && not is_last_stage
+                # Send dx to NEXT stage;
+                ################
+                else:
+                    next_rank = self.stage_manager.get_next_rank()
+                    input_tensor_grad = self.send_backward_buffer[model_chunk_id].pop(0)
+                    send_handles = self.comm.send_backward(input_tensor_grad, next_rank)
+                    return send_handles
+
+    def forward_step(
+        self,
+        model_chunk: Union[ModuleList, Module],
+        model_chunk_id: int,
+        input_obj: Optional[dict],
+        criterion: Callable,
+        accum_loss: Optional[torch.Tensor] = None,
+        outputs: Optional[List[Any]] = None,
+    ) -> Union[torch.Tensor, dict]:
+        """Forward one step of the pipeline
+        Args:
+            model_chunk (ModuleList or Module): Model Chunk to be run;
+            model_chunk_id (int): The current model chunk idx;
+            input_obj (Optional[dict]): x;
+            criterion (Callable): loss function;
+            accum_loss (Optional[torch.Tensor], optional): Accumulated loss. Defaults to None.
+            outputs (Optional[List[Any]], optional): List to store the output of the last stage (final output). Defaults to None.
+
+        Returns:
+            Union[torch.Tensor, dict]: The intermediate output (dict) of the current stage. If it is the last stage, the output is the loss (Tensor).
+        """
+        # Load input ids, attention mask and labels
+        # micro_batch = self.load_micro_batch(model_chunk_id=model_chunk_id)
+
+        # for the first stage, input_obj is None
+        # for other stages, input_obj is the output of the previous/next stage containing hidden_states etc.
+        # Only attention_mask from micro_batch is used
+
+        with self.stage_manager.switch_model_chunk_id(model_chunk_id):
+            # fwd calculate
+            output_obj = model_chunk[model_chunk_id](input_obj)
+            # last layer in model
+            if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
+                loss = criterion(output_obj) / self.num_microbatch
+                if accum_loss is not None:
+                    accum_loss.add_(loss.detach())
+                if outputs is not None:
+                    outputs.append(tree_map(detach, output_obj))
+                # print(f"accum_loss {accum_loss}; outputs {len(outputs)}; model_chunk_id {model_chunk_id}")
+                return loss
+            else:
+                return output_obj
```

Comment:
Where do you release the storage of output tensors? They are not used during backward pass.

Reply:
> Where do you release the storage of output tensors? They are not used during backward pass.

Use a buffer to append and pop it;
We append y after fwd_step;
<img width="545" alt="c0e84b258671cd016ab35d844a28c85" src="https://github.com/user-attachments/assets/8a748c55-28d5-45b2-8c0c-e929e0a73457">

Then we pop y before bwd step;
<img width="531" alt="21716b955437bf6770a4603beb2ebfb" src="https://github.com/user-attachments/assets/132f260f-3186-435f-a6d4-b1865c7ba08f">

Reply:
I know, but actually we don't need the data of output during backward pass, right? What we save in buffer should be a tensor in graph without concrete data.

Reply:
The data can be released after sending.

Reply:
> The data can be released after sending.

fixed in https://github.com/hpcaitech/ColossalAI/pull/6034/commits/ab643c9af74a57d7e5fcdbf38c31b596db819a5b